tests/test_save.py

```diff
@@ -0,0 +1,250 @@
+    assert(dc == state)
+
+    #Save the agent and deserialize
+    agent_saved = dill.dumps(agent)
+    agent_deserialized = dill.loads(agent_saved)
+
+    #check that the agent is same before and after serializing
+    areAgentsSame(agent, agent_deserialized)
+
+    #planner is different right now bc of lambda
+    # assert(agent.__dict__['planner'].__dict__ == agent_deserialized.__dict__['planner'].__dict__)
+
+    #request both (train changes state?)
+    SAI1 = agent.request(state, reward = 1)
+    SAI2 = agent_deserialized.request(state, reward = 1)
+
+    assert(SAI1 == SAI2)
+
+"""
+This module will test initial creation and do simple train/request tests on
+all possible configurations of agent and learner types. The goal is mainly to
+be a break glass sanity check, note to test behavior.
+"""
+
+logging.disable()
+module_path = dirname(__file__)
+
+
+def train_A_req_A(agent):
+    add_set = json.load(open(join(module_path, 'data_for_test.json'), 'r'))['addition']
+
+    for data in add_set[:-1]:
+        sai = Sai(
+            selection=data["selection"],
+            action=data["action"],
+            inputs=data["inputs"],
+        )
+
+        data['sai'] = sai
+        del data['selection']
+        del data['action']
+        del data['inputs']
+
+        agent.train(**data)
+
+    return agent.request(add_set[-1]['state']), add_set[-1]
+
+
+def req_A(agent):
+    add_set = json.load(open(join(module_path, 'data_for_test.json'), 'r'))['addition']
+    return agent.request(add_set[-1]['state'])
+
+
+@pytest.mark.parametrize("when", ['decision_tree', 'cobweb', 'trestle'])
+@pytest.mark.parametrize("where", ['version_space',
+                                   'fastmostspecific',
+                                   'mostspecific'
+                                   # ,'stateresponselearner',
+                                   # 'relationallearner',
+                                   #'specifictogeneral'
+                                   ])
+@pytest.mark.parametrize("which", ['proportioncorrect', 'totalcorrect'])
+@pytest.mark.parametrize("choice", ['first', 'mostparsimonious', 'all', 'random'])
+def test_numbert_configs(when, where, which, choice):
+    function_set = ["RipFloatValue", "Add", "Subtract", "Multiply", "Divide"]
+    feature_set = ["Equals"]
+    agent = ModularAgent(feature_set, function_set,
+                         when_learner=when,
+                         where_learner=where,
+                         planner='numbert',
+                         heuristic_learner=which,
+                         explanation_choice=choice,
+                         search_depth=2)
+    resp = req_A(agent)
+    assert resp == {}
+    
+    resp, ans = train_A_req_A(agent)
+    assert resp['selection'] == ans['selection']
+    assert resp['action'] == ans['action']
+    assert str(int(resp['inputs']['value'])) == ans['inputs']['value']
+
+    #check that copy is the same
+    dc = copy.deepcopy(resp)
+    assert(dc == resp)
```

Comment:
This check seems like an unnecessary test since its just confirming the behavior of the external deepcopy library working on a built in dictionary type.